# Project: Market Basket Analysis (Market Basket Optimisation Dataset)

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#training apriori in dataset">Training Apriori on Dataset and Visualising the Result</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

>### **Overview**
>To complete my Data Analysis project I am using Market Basket Optimisation Dataset. 

>This data set contains information about 7500 transaction collected from marketplace. It consist of 20 columns item. 

>#### **Question that can analysed from this data set**
> If the rules are min_support = 0.004, min_confidence = 0.2, min_lift = 3, min_length = 2, People who bought what item also bought what item? and what the value of support?

In [2]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [3]:
# Data Preprocessing
dataset = pd.read_csv('Market_Basket_Optimisation.csv', header = None)
transactions = []
for i in range(0, 7501):
    transactions.append([str(dataset.values[i,j]) for j in range(0, 20)])

In [18]:
dataset.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [20]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 20 columns):
0     7501 non-null object
1     5747 non-null object
2     4389 non-null object
3     3345 non-null object
4     2529 non-null object
5     1864 non-null object
6     1369 non-null object
7     981 non-null object
8     654 non-null object
9     395 non-null object
10    256 non-null object
11    154 non-null object
12    87 non-null object
13    47 non-null object
14    25 non-null object
15    8 non-null object
16    4 non-null object
17    4 non-null object
18    3 non-null object
19    1 non-null object
dtypes: object(20)
memory usage: 1.1+ MB


<a id='training apriori in dataset'></a>
## Training Apriori and Visualising the Result

In [34]:
# Training Apriori on the dataset
from apyori import apriori
rules = apriori(transactions, min_support = 0.004, min_confidence = 0.2, min_lift = 3, min_length = 2)

#we can change min_support, min_confidence, min_lift, and min_length.

In [35]:
# Visualising the results
results = list(rules)

In [36]:
results=pd.DataFrame(results)
results

items   support  \
0                              (light cream, chicken)  0.004533   
1                    (mushroom cream sauce, escalope)  0.005733   
2                                   (pasta, escalope)  0.005866   
3                        (herb & pepper, ground beef)  0.015998   
4                         (tomato sauce, ground beef)  0.005333   
5                      (olive oil, whole wheat pasta)  0.007999   
6                                     (shrimp, pasta)  0.005066   
7                         (light cream, chicken, nan)  0.004533   
8              (shrimp, frozen vegetables, chocolate)  0.005333   
9               (cooking oil, ground beef, spaghetti)  0.004799   
10                 (eggs, herb & pepper, ground beef)  0.004133   
11              (mushroom cream sauce, nan, escalope)  0.005733   
12                             (pasta, nan, escalope)  0.005866   
13        (frozen vegetables, ground beef, spaghetti)  0.008666   
14               (olive oil, milk, frozen vegetables)  0.004799   
15                (milk, frozen vegetables, tomatoes)  0.004133   
16         (mineral water, shrimp, frozen vegetables)  0.007199   
17          (olive oil, frozen vegetables, spaghetti)  0.005733   
18             (shrimp, frozen vegetables, spaghetti)  0.005999   
19           (frozen vegetables, tomatoes, spaghetti)  0.006666   
20            (ground beef, grated cheese, spaghetti)  0.005333   
21        (mineral water, herb & pepper, ground beef)  0.006666   
22                  (nan, herb & pepper, ground beef)  0.015998   
23            (herb & pepper, ground beef, spaghetti)  0.006399   
24                     (olive oil, milk, ground beef)  0.004933   
25                   (tomato sauce, nan, ground beef)  0.005333   
26                   (shrimp, ground beef, spaghetti)  0.005999   
27                       (olive oil, milk, spaghetti)  0.007199   
28                   (olive oil, mineral water, soup)  0.005199   
29                (whole wheat pasta, olive oil, nan)  0.007999   
30                               (shrimp, pasta, nan)  0.005066   
31                   (olive oil, pancakes, spaghetti)  0.005066   
32                   (olive oil, tomatoes, spaghetti)  0.004399   
33        (shrimp, nan, frozen vegetables, chocolate)  0.005333   
34         (cooking oil, nan, ground beef, spaghetti)  0.004799   
35            (eggs, herb & pepper, ground beef, nan)  0.004133   
36  (mineral water, frozen vegetables, ground beef...  0.004399   
37   (nan, frozen vegetables, ground beef, spaghetti)  0.008666   
38  (mineral water, milk, frozen vegetables, spagh...  0.004533   
39          (olive oil, milk, frozen vegetables, nan)  0.004799   
40           (milk, frozen vegetables, tomatoes, nan)  0.004133   
41    (mineral water, shrimp, nan, frozen vegetables)  0.007199   
42     (olive oil, nan, frozen vegetables, spaghetti)  0.005733   
43        (shrimp, nan, frozen vegetables, spaghetti)  0.005999   
44      (nan, frozen vegetables, tomatoes, spaghetti)  0.006666   
45       (nan, ground beef, grated cheese, spaghetti)  0.005333   
46   (mineral water, nan, herb & pepper, ground beef)  0.006666   
47       (nan, herb & pepper, ground beef, spaghetti)  0.006399   
48                (olive oil, milk, ground beef, nan)  0.004933   
49              (shrimp, nan, ground beef, spaghetti)  0.005999   
50                  (olive oil, milk, spaghetti, nan)  0.007199   
51              (olive oil, nan, mineral water, soup)  0.005199   
52              (olive oil, pancakes, nan, spaghetti)  0.005066   
53              (olive oil, nan, tomatoes, spaghetti)  0.004399   
54  (ground beef, mineral water, nan, frozen veget...  0.004399   
55  (milk, mineral water, nan, frozen vegetables, ...  0.004533   

                                   ordered_statistics  
0   [((light cream), (chicken), 0.2905982905982905...  
1   [((mushroom cream sauce), (escalope), 0.300699...  
2   [((pasta), (escalope), 0.3728813559322034, 4.7...  
3   

In [38]:
results.sort_values('support',ascending=False)

items   support  \
22                  (nan, herb & pepper, ground beef)  0.015998   
3                        (herb & pepper, ground beef)  0.015998   
13        (frozen vegetables, ground beef, spaghetti)  0.008666   
37   (nan, frozen vegetables, ground beef, spaghetti)  0.008666   
5                      (olive oil, whole wheat pasta)  0.007999   
29                (whole wheat pasta, olive oil, nan)  0.007999   
27                       (olive oil, milk, spaghetti)  0.007199   
41    (mineral water, shrimp, nan, frozen vegetables)  0.007199   
50                  (olive oil, milk, spaghetti, nan)  0.007199   
16         (mineral water, shrimp, frozen vegetables)  0.007199   
46   (mineral water, nan, herb & pepper, ground beef)  0.006666   
19           (frozen vegetables, tomatoes, spaghetti)  0.006666   
44      (nan, frozen vegetables, tomatoes, spaghetti)  0.006666   
21        (mineral water, herb & pepper, ground beef)  0.006666   
47       (nan, herb & pepper, ground beef, spaghetti)  0.006399   
23            (herb & pepper, ground beef, spaghetti)  0.006399   
49              (shrimp, nan, ground beef, spaghetti)  0.005999   
26                   (shrimp, ground beef, spaghetti)  0.005999   
43        (shrimp, nan, frozen vegetables, spaghetti)  0.005999   
18             (shrimp, frozen vegetables, spaghetti)  0.005999   
12                             (pasta, nan, escalope)  0.005866   
2                                   (pasta, escalope)  0.005866   
17          (olive oil, frozen vegetables, spaghetti)  0.005733   
42     (olive oil, nan, frozen vegetables, spaghetti)  0.005733   
11              (mushroom cream sauce, nan, escalope)  0.005733   
1                    (mushroom cream sauce, escalope)  0.005733   
8              (shrimp, frozen vegetables, chocolate)  0.005333   
25                   (tomato sauce, nan, ground beef)  0.005333   
45       (nan, ground beef, grated cheese, spaghetti)  0.005333   
4                         (tomato sauce, ground beef)  0.005333   
33        (shrimp, nan, frozen vegetables, chocolate)  0.005333   
20            (ground beef, grated cheese, spaghetti)  0.005333   
28                   (olive oil, mineral water, soup)  0.005199   
51              (olive oil, nan, mineral water, soup)  0.005199   
30                               (shrimp, pasta, nan)  0.005066   
52              (olive oil, pancakes, nan, spaghetti)  0.005066   
6                                     (shrimp, pasta)  0.005066   
31                   (olive oil, pancakes, spaghetti)  0.005066   
48                (olive oil, milk, ground beef, nan)  0.004933   
24                     (olive oil, milk, ground beef)  0.004933   
34         (cooking oil, nan, ground beef, spaghetti)  0.004799   
9               (cooking oil, ground beef, spaghetti)  0.004799   
39          (olive oil, milk, frozen vegetables, nan)  0.004799   
14               (olive oil, milk, frozen vegetables)  0.004799   
0                              (light cream, chicken)  0.004533   
38  (mineral water, milk, frozen vegetables, spagh...  0.004533   
7                         (light cream, chicken, nan)  0.004533   
55  (milk, mineral water, nan, frozen vegetables, ...  0.004533   
36  (mineral water, frozen vegetables, ground beef...  0.004399   
32                   (olive oil, tomatoes, spaghetti)  0.004399   
53              (olive oil, nan, tomatoes, spaghetti)  0.004399   
54  (ground beef, mineral water, nan, frozen veget...  0.004399   
40           (milk, frozen vegetables, tomatoes, nan)  0.004133   
35            (eggs, herb & pepper, ground beef, nan)  0.004133   
15                (milk, frozen vegetables, tomatoes)  0.004133   
10                 (eggs, herb & pepper, ground beef)  0.004133   

                                   ordered_statistics  
22  [((nan, herb & pepper), (ground beef), 0.32345...  
3   [((herb & pepper), (ground beef), 0.3234501347...  
13  [((frozen vegetables, spaghetti), (ground beef...  
37  

<a id='conclusions'></a>
## Conclusions

As we can see from the table above, 

1. People who bought herb & pepper also bought ground beef with support 0.015998
2. People who bought frozen vegetables also bought ground beef and spaghetti with support 0.008666
3. People who bought olive oil also bought whole wheat pasta with support 0.007999
4. etc